pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java

```diff
@@ -82,14 +103,220 @@ public void setTableConfigFileName(String tableConfigFileName) {
+      Thread.sleep(DEFAULT_SLEEP_INTERVAL_MS);
+    }
+
+    LOGGER.info("Successfully verified segment {} and its current status is {}.", _segmentName, state);
+    return true;
+  }
+
+  /**
+   * Deletes the segment for the given segment name and table name.
+   * @return true if delete successful, else false.
+   */
+  private boolean deleteSegment() {
+    try {
+      TableConfig tableConfig = JsonUtils.fileToObject(new File(_tableConfigFileName), TableConfig.class);
+      _tableName = tableConfig.getTableName();
+
+      ControllerTest.sendDeleteRequest(ControllerRequestURLBuilder.baseUrl(ClusterDescriptor.CONTROLLER_URL)
+          .forSegmentDelete(_tableName, _segmentName));
+      return verifySegmentDeleted();
+    } catch (Exception e) {
+      LOGGER.error("Request to delete the segment {} for the table {} failed.", _segmentName, _tableName, e);
+      return false;
+    }
+  }
+
+  /**
+   * Verify given table name and segment name deleted from the controller.
+   * @return true if no segment found, else false.
+   * @throws IOException
+   * @throws InterruptedException
+   */
+  private boolean verifySegmentDeleted()
+      throws IOException, InterruptedException {
+    long startTime = System.currentTimeMillis();
+    while (getCountForSegmentName() > 0) {
+      if ((System.currentTimeMillis() - startTime) > DEFAULT_MAX_SLEEP_TIME_MS) {
+        LOGGER.error("Delete segment verification failed, count is greater than zero after max wait time {} ms.",
+            DEFAULT_MAX_SLEEP_TIME_MS);
+        return false;
+      }
+      LOGGER.warn("Delete segment verification count greater than zero, will retry after {} ms.",
+          DEFAULT_SLEEP_INTERVAL_MS);
+      Thread.sleep(DEFAULT_SLEEP_INTERVAL_MS);
+    }
+
+    LOGGER.info("Successfully delete the segment {} for the table {}.", _segmentName, _tableName);
+    return true;
+  }
+
+  /**
+   * Retrieve external view for the given table name.
+   * @return TableViews.TableView of OFFLINE and REALTIME segments.
+   */
+  private TableViews.TableView getExternalViewForTable()
+      throws IOException {
+    return JsonUtils.stringToObject(ControllerTest.sendGetRequest(
+        ControllerRequestURLBuilder.baseUrl(ClusterDescriptor.CONTROLLER_URL).forTableExternalView(_tableName)),
+        TableViews.TableView.class);
+  }
+
+  /**
+   * Retrieve the number of segments for both OFFLINE and REALTIME which are in state matching the parameter.
+   * @param state of the segment to be verified in the controller.
+   * @return count for OFFLINE and REALTIME segments.
+   */
+  private long getSegmentCountInState(String state)
+      throws IOException {
+    long offlineSegmentCount =
+        getExternalViewForTable().offline != null ? getExternalViewForTable().offline.entrySet().stream()
+            .filter(k -> k.getKey().equalsIgnoreCase(_segmentName)).filter(v -> v.getValue().values().contains(state))
```

Comment:
segment names are case sensitive. Please do not ignore case.
Also, if there are two replicas, and one of them is ONLINE and the other is not, how will this work?
Instead of counting the number, can we get the state of all replicas of the segment and return true ONLY if they are in the intended state?

Reply:
Considering the below sample table's external view, where replicas are 2, the expectation is both `Server_hostname1_8001` and `Server_hostname2_8001` should in `ONLINE` else if its should be derived segment is `OFFLINE`?

**Sample Table External view:**
```{
  "id" : "table_OFFLINE",
  "simpleFields" : {
    "BUCKET_SIZE" : "0",
    "INSTANCE_GROUP_TAG" : "table_OFFLINE",
    "MAX_PARTITIONS_PER_INSTANCE" : "1",
    "NUM_PARTITIONS" : "1",
    "REBALANCE_MODE" : "CUSTOMIZED",
    "REPLICAS" : "2"
  },
  "mapFields" : {
    "account_summary_additive_daily_0" : {
      "Server_hostname1_8001" : "ONLINE",
      "Server_hostname2_8001" : "ONLINE"
    }
  },
  "listFields" : {
  }
}```

Reply:
Yes, the expectation is that all replicas reach ONLINE. Some replicas may take more time to get there than others, so we should wait until all replicas reach ONLINE state. If any replica reaches ERROR state, we can bail out early with a failure.

Reply:
Fixed